greenplumpython/func.py

```diff
@@ -0,0 +1,36 @@
+from typing import Callable, Iterable, Optional
+from uuid import uuid4
+
+from .db import Database
+from .expr import Expr
+from .table import Table
+
+
+class FunctionCall(Expr):
+    def __init__(
+        self, func_name: str, db: Database, args: Iterable[Expr] = [], as_name: Optional[str] = None
+    ) -> None:
+        super().__init__(as_name)
+        self._func_name = func_name
+        self._args = args
+        self._db = db
+
+    def __str__(self) -> str:
+        args_string = ",".join([str(arg) for arg in self._args]) if any(self._args) else ""
+        return f"{self._func_name}({args_string})"
+
+    def to_table(self) -> Table:
+        as_string = f"AS {self._as_name}" if self._as_name is not None else ""
+        ret_table = Table(f"SELECT * FROM {str(self)} {as_string}", db=self._db)
+        return Table(f"SELECT * FROM {ret_table.name}", parents=[ret_table], db=self._db)
```

Comment:
Good question! This is to support functions whose return type is composite. 

For example, the user has such a function `func`. When the user calls it, they would typically need to unwrap the result of the UDF into individual colummns. This can be typically done with `.*`, like this:

```sql
SELECT (func(t.a)).* FROM t
```

However, this will lead to severe performance issue because the function will be called multiple times on GPDB 6X.

One way to workaround is to wrap the function call with a CTE:
```sql
With ret_table AS (
    SELECT func(t.a) AS ret FROM t
)
SELECT ret.* FROM ret_table
```

This is why the old `gpapply()` calls UDF in such an awkward way: https://github.com/greenplum-db/GreenplumPython/blob/10830a41d0e4b2ad2f96a8449fa07848ab29391b/greenplumpython/core/gpapply.py#L21